src/v/security/krb5.h

```diff
@@ -10,82 +10,100 @@
 
 #pragma once
 
+#include "outcome.h"
+#include "seastarx.h"
+
+#include <seastar/core/sstring.hh>
+
 #include <krb5/krb5.h>
 
+#include <optional>
 #include <string_view>
 
 namespace security::krb5 {
 
-class krb5_context_view {
+class context_view {
 public:
-    explicit krb5_context_view(const ::krb5_context* ctx)
+    explicit context_view(::krb5_context ctx)
       : _ctx(ctx) {}
 
-    ::krb5_context operator*() const { return *_ctx; }
-    explicit operator const ::krb5_context*() { return _ctx; }
+    ::krb5_context operator*() const { return _ctx; }
 
 private:
-    const ::krb5_context* _ctx;
+    ::krb5_context _ctx;
 };
 
-class krb5_context {
-public:
-    krb5_context() = default;
-    krb5_context(krb5_context&&) = delete;
-    krb5_context(const krb5_context&) = delete;
-    krb5_context operator=(krb5_context&&) = delete;
-    krb5_context operator=(const krb5_context&) = delete;
-
-    ~krb5_context() noexcept { reset(); }
-
-    void reset() noexcept {
-        if (_ctx != nullptr) {
-            ::krb5_free_context(_ctx);
-            _ctx = nullptr;
+namespace impl {
+class error_message {
+    struct deleter {
+        std::optional<context_view> ctx;
+        void operator()(const char* realm) const {
+            ::krb5_free_error_message(ctx ? **ctx : nullptr, realm);
```

Comment:
This looks more complicated than it is I think, due to the operator* mentioned previously.

Would it be clearer to let the null propagate from a failed `::krb5_init_context` and just use that everywhere? The library seems to do the right thing for a null context, which is what we end up passing here anyway, but now there are two code paths:
* Here to see if the optional is engaged.
* `error_message` constructor implementation and call-site (could just construct with a view of a null context?)

Reply:
This is purely for the edge case of when `krb5_init_context` fails.  I guess I can just create a null view and pass that along.